src/main/java/com/linecorp/armeria/client/RemoteInvokerFactory.java

```diff
@@ -219,4 +233,85 @@ private void close0() {
             eventLoopGroup.shutdownGracefully().syncUninterruptibly();
         }
     }
+
+    // TODO(trustin): Remove this once the upstream fixes the problem:
+    //                https://github.com/netty/netty/issues/5179
+    private static final class DnsAddressResolverGroup5179 extends DnsAddressResolverGroup {
+
+        private final Map<String, Promise<InetAddress>> resolveMap = new ConcurrentHashMap<>();
+        private final Map<String, Promise<List<InetAddress>>> resolveAllMap = new ConcurrentHashMap<>();
+
+        DnsAddressResolverGroup5179(Class<? extends DatagramChannel> channelType,
+                                    DnsServerAddresses nameServerAddresses) {
+            super(channelType, nameServerAddresses);
+        }
+
+        @Override
+        protected AddressResolver<InetSocketAddress> newResolver(
+                EventLoop eventLoop, ChannelFactory<? extends DatagramChannel> channelFactory,
+                InetSocketAddress localAddress, DnsServerAddresses nameServerAddresses) throws Exception {
+
+            final InetNameResolver delegate = new DnsNameResolverBuilder(eventLoop)
+                    .channelFactory(channelFactory)
+                    .localAddress(localAddress)
+                    .nameServerAddresses(nameServerAddresses)
+                    .build();
+
+            return new DnsNameResolver5179(eventLoop, delegate).asAddressResolver();
+        }
+
+        private final class DnsNameResolver5179 extends InetNameResolver {
+
+            private final InetNameResolver delegate;
+
+            DnsNameResolver5179(EventExecutor executor, InetNameResolver delegate) {
+                super(executor);
+                this.delegate = delegate;
+            }
+
+            @Override
+            protected void doResolve(String inetHost, Promise<InetAddress> promise) throws Exception {
+                doResolve0(delegate::resolve, resolveMap, inetHost, promise);
+            }
+
+            @Override
+            protected void doResolveAll(String inetHost, Promise<List<InetAddress>> promise) throws Exception {
+                doResolve0(delegate::resolveAll, resolveAllMap, inetHost, promise);
+            }
+
+            private <T> void doResolve0(
+                    BiConsumer<String, Promise<T>> resolver, Map<String, Promise<T>> resolveMap,
+                    String inetHost, Promise<T> promise) throws Exception {
+
+                final Promise<T> earlyPromise = resolveMap.putIfAbsent(inetHost, promise);
+                if (earlyPromise != null) {
+                    // Name resolution for the specified inetHost is in progress already.
+                    if (earlyPromise.isDone()) {
```

Comment:
If I get it right, this branch exists for the sole purpose of not instanciating a listener if the earlyPromise is done. But isn't this statistically unlikely? Is it really worth adding a branch in the code? Just asking.